flow/raster_cache.cc

```diff
@@ -27,6 +27,18 @@ RasterCacheResult::RasterCacheResult(sk_sp<SkImage> image,
                                      const SkRect& logical_rect)
     : image_(std::move(image)), logical_rect_(logical_rect) {}
 
+void RasterCacheResult::drawSnapshot(SkCanvas& canvas) const {
+  SkAutoCanvasRestore auto_restore(&canvas, true);
+  canvas.resetMatrix();
+  SkIRect dev_clip = canvas.getDeviceClipBounds();
+  if (dev_clip.width() == logical_rect_.width() &&
+      dev_clip.height() == logical_rect_.height()) {
+    canvas.drawImage(image_, dev_clip.left(), dev_clip.top(), nullptr);
+  } else {
+    canvas.drawImageRect(image_, SkRect::Make(dev_clip), nullptr);
```

Comment:
I'm not sure why we want to draw a scaled image here instead of clipping.

For example, if the old clip is (0, 0, 100, 100) and the new clip is (0, 50, 100, 100), are we going to scale the image by (1.0, 0.5)?

Reply:
Yes, and we also scale it up. If we don't scale it, then the child may grow or shrink while its blurred background stays the same size.

I test the case in the sample app mentioned in the benchmark comment, but I am open to other interpretations of how to handle that case. The current implementation allows the developer to choose static/dynamic based on whether they like the stretching effect or not.

I could choose to recache in this case, but any change of size in the child would disrupt the caching.

I could choose to leave it always the same size (and center it over the new clipped area?), but then the child would be growing with the blur not growing.

If the developer wants the blur to remain a constant size and not be stretched, they can put a statically sized blank widget in the child to keep the clip size stable while other parts of the child move/scale/rotate etc.

If the developer wants the blur to change size, they can set the filter widget back to dynamic to drop the caching.

Reply:
Since BackdropFilter's size is not decided by its child, but rather by its parent clip, I don't think a changing child should change our snapshot? On the other hand, if the parent clip does change, I do think that invalidate the cache is more proper.

Reply:
In the version in my latest commit this code now blurs the entire background and then renders it through the clip so the developer can modify the clip and animate the child all they want and we will render the appropriate backdrop as without caching, though the layering is different as discussed in a different comment thread.

Reply:
(To be clear, no scaling of the cached backdrop happens now at all.)